### DISCUSSION TITLE
Flux dry-run fails with things that depends on other

### DISCUSSION THREAD

**Discussion by midestefanis** [fluxcd/flux2#3108] [Q&A]
`Kustomization reconciliation failed: VirtualNode/access-proxy dry-run failed, reason: failed to find matching mesh for namespace: ***, expecting 1 but found 0, error: admission webhook "mvirtualnode.appmesh.k8s.aws" denied the request: failed to find matching mesh for namespace: ***, expecting 1 but found 0`

The problem is that the dry-run fails because the same kustomization first installs "Mesh" and then "VirtualNode/Service" but the dry-run doesn't know about it. Any suggestion or I have no choice but to do it through 'kustomization' and 'dependsOn'?

**Comment by kingdonb** [comment]
> have no choice but to do it through 'kustomization' and 'dependsOn'

The latest release of Flux includes new logic to order resources by pattern matching, so that "ClusterClass" is installed before "Cluster" for example. Flux does not / cannot get all of these combinations right, some resources must be created before others. The `dependsOn` pattern is the one that users are expected to take advantage of here, I'm not sure there's any other option for now. Since the `Mesh` and `VirtualNode` resources are not obviously related here, I don't think there's any rule that we could apply to guarantee they are applied in the right order.

We could potentially add them to the list in `fluxcd/flux2/pkg/ssa` but as I said the last time this came up, there is a concern that list cannot grow forever.

Here is that list in the source code:
https://github.com/fluxcd/pkg/blob/884d4b935471b224047e50d796cf71e0e9582255/ssa/sort.go#L35

It's full of kubernetes-native objects, cluster-api objects, (I'm not sure if objects in the AppMesh API would be considered for inclusion or not.)

**Reply by rombert** [reply]
Would it be an option to have users extend this somehow?

**Reply by stefanprodan** [reply]
> Would it be an option to have users extend this somehow?

Not really, what @kingdonb proposed is not going to have any effect for admission controllers. That list in `sort.go` does nothing, is more of a documentation thing.

**Comment by stefanprodan** [comment]
Here is an example of how to structure a repo for AppMesh using `dependsOn` https://github.com/stefanprodan/gitops-appmesh/tree/main/infrastructure
